File: python_scripts/MakeManuscriptFigures.py
"""
MakeManuscriptFigures - one-shot orchestrator that runs every manuscript
figure generator in sequence and prints a PASS/FAIL summary.

Python counterpart of matlab_scripts/MakeManuscriptFigures.m (todo #104).

Figure -> generator map (mirrors the MATLAB orchestrator):
    characterization_combined.pdf      -> PlotCharacterizationData_Cell15.py (see NOTE below)
    NextBMS_ReferencePerformanceCycle  -> PlotReferencePerformanceCycleFigure_Cell68.py
    LiStrippingMethods.pdf             -> PlotLiStrippingMethods_Cell35.py
    Anode.pdf / Cathode.pdf            -> extractOCPLines.py
    <cell>_CharacterizationOCV/EIS.pdf -> PlotCharacterizationResults.py (Cell_4/Cell_6/Cell_15)
    CyclicAgeing.pdf / CalendarAgeing  -> PlotAgeingCombinedOverview.py
    <cell>_Summary.pdf                 -> PlotCellSummary.py (example cell)

NOTE (documented decision): `python_scripts/PlotCharacterizationData_Cell15.py` is a
generic per-file debug plotter (todo #006) and does NOT yet reproduce the
`characterization_combined.pdf` figure that the MATLAB
`PlotCharacterizationData_Cell15.m` produces (arrows, journal styling,
single 5-phase combined figure). It is still called here for parity of
orchestration order, but this stage is expected to produce different
(debug-only) output until #006 is completed - flagged as PARTIAL below.

Author: GitHub Copilot (for Robinson Medina / Feye Hoekstra)
Date: 2026-08-24

Compliance: R-001, R-013, R-016, R-022.
"""

import importlib
import shutil
import sys
import time
import traceback
from pathlib import Path

_SCRIPT_DIR = Path(__file__).resolve().parent
if str(_SCRIPT_DIR) not in sys.path:
    sys.path.append(str(_SCRIPT_DIR))

_FUNCTIONS_DIR = _SCRIPT_DIR.parent / 'Functions'
if not _FUNCTIONS_DIR.exists():
    _FUNCTIONS_DIR = _SCRIPT_DIR.parent.parent / 'Functions'
if str(_FUNCTIONS_DIR) not in sys.path:
    sys.path.append(str(_FUNCTIONS_DIR))
from get_figure_output_dir import get_figure_output_dir  # noqa: E402

# Own R-022 folder that collects every manuscript figure this driver produces.
_MANUSCRIPT_DIR = get_figure_output_dir('MakeManuscriptFigures')
_FIGURES_PYTHON_ROOT = _MANUSCRIPT_DIR.parent


def _consolidate(stems, since_ts):
    """Copy files each generator stem produced/updated during this stage into the
    MakeManuscriptFigures folder, so the manuscript figures live in one place."""

    for stem in stems:
        src_dir = _FIGURES_PYTHON_ROOT / stem
        if not src_dir.is_dir():
            continue
        for src in src_dir.iterdir():
            if not src.is_file() or src.suffix.lower() not in ('.pdf', '.png'):
                continue
            if src.stat().st_mtime + 1e-3 < since_ts:
                continue  # untouched by this stage
            try:
                shutil.copy2(src, _MANUSCRIPT_DIR / src.name)
            except OSError as copy_exc:  # noqa: PERF203 - report, do not abort
                print(f'[WARN] could not copy {src.name}: {copy_exc}')


def _run_stage(label, fn, *args, stems=(), **kwargs):
    """Run one stage, catching and reporting failures without stopping the run."""

    print('\n' + '#' * 60)
    print(f'STAGE: {label}')
    print('#' * 60)
    since_ts = time.time()
    try:
        fn(*args, **kwargs)
        _consolidate(stems, since_ts)
        return True, None
    except Exception as exc:  # noqa: BLE001 - mirrors MATLAB's per-stage try/catch
        print(f'[FAIL] {label}: {exc}')
        traceback.print_exc()
        return False, str(exc)


def main():
    """Run every manuscript-figure generator in sequence."""

    results = []

    ok, err = _run_stage(
        'PlotCharacterizationData_Cell15 (characterization_combined.pdf - PARTIAL, see #006)',
        lambda: importlib.import_module('PlotCharacterizationData_Cell15').main(),
        stems=('PlotCharacterizationData_Cell15',))
    results.append(('PlotCharacterizationData_Cell15', ok, err))

    ok, err = _run_stage(
        'PlotReferencePerformanceCycleFigure_Cell68 (NextBMS_ReferencePerformanceCycle.pdf)',
        lambda: importlib.import_module('PlotReferencePerformanceCycleFigure_Cell68').main(),
        stems=('PlotReferencePerformanceCycleFigure_Cell68',))
    results.append(('PlotReferencePerformanceCycleFigure_Cell68', ok, err))

    ok, err = _run_stage(
        'PlotLiStrippingMethods_Cell35 (LiStrippingMethods.pdf)',
        lambda: importlib.import_module('PlotLiStrippingMethods_Cell35').main(),
        stems=('PlotLiStrippingMethods_Cell35',))
    results.append(('PlotLiStrippingMethods_Cell35', ok, err))

    ok, err = _run_stage(
        'extractOCPLines (Anode.pdf / Cathode.pdf)',
        lambda: importlib.import_module('extractOCPLines').main(),
        stems=('extractOCPLines',))
    results.append(('extractOCPLines', ok, err))

    ok, err = _run_stage(
        'PlotCharacterizationResults (<cell>_CharacterizationOCV/EIS.pdf x3)',
        lambda: [importlib.import_module('PlotCharacterizationResults').main(cell_num=c)
                 for c in ('Cell_4', 'Cell_6', 'Cell_15')],
        stems=('PlotCharacterizationResults',))
    results.append(('PlotCharacterizationResults', ok, err))

    ok, err = _run_stage(
        'PlotAgeingCombinedOverview (CyclicAgeing.pdf / CalendarAgeing.pdf)',
        lambda: importlib.import_module('PlotAgeingCombinedOverview').main(),
        stems=('PlotAgeingCombinedOverview',))
    results.append(('PlotAgeingCombinedOverview', ok, err))

    ok, err = _run_stage(
        'PlotCellSummary (example cell: Cell_35_Summary.pdf)',
        lambda: importlib.import_module('PlotCellSummary').main(cell_num='Cell_35'),
        stems=('PlotCellSummary',))
    results.append(('PlotCellSummary', ok, err))

    print('\n' + '=' * 60)
    print('MakeManuscriptFigures: summary')
    print('=' * 60)
    n_pass = sum(1 for _, ok, _ in results if ok)
    for name, ok, err in results:
        status = 'PASS' if ok else f'FAIL ({err})'
        print(f'  {name:<45s} {status}')
    print(f'\n{n_pass}/{len(results)} stages passed.')
    print(f'Manuscript figures collected in: {_MANUSCRIPT_DIR}')


if __name__ == '__main__':
    main()

File: matlab_scripts/MakeManuscriptFigures.m
%% MakeManuscriptFigures.m
% Summary: Runs the MATLAB generators needed to reproduce the code-produced
%          manuscript figures. Each stage is isolated, reports PASS or FAIL,
%          and leaves generated outputs in the generator's R-022 directory.
%
% Usage: Run from MATLAB or with matlab -batch "run('MakeManuscriptFigures.m')".
%        The driver continues after a failed stage so the final summary shows
%        which figures were generated and which still need attention.
%
% Outputs: A terminal summary for all stages. Figure and CSV outputs are written
%          by the called generator scripts, not by this driver.
%
% Authors: Feye Hoekstra, GitHub Copilot.
% Dependency files: matlab_scripts/PlotCharacterizationData_Cell15.m,
%                   matlab_scripts/PlotReferencePerformanceCycleFigure_Cell68.m,
%                   matlab_scripts/PlotLiStrippingMethods_Cell35.m,
%                   matlab_scripts/extractOCPLines.m,
%                   matlab_scripts/PlotCharacterizationResults.m,
%                   matlab_scripts/ExtractAgeingData.m,
%                   matlab_scripts/PlotEISData.m,
%                   matlab_scripts/PlotAgeingCombinedOverview.m,
%                   matlab_scripts/PlotCellSummary.m,
%                   Functions/getFigureOutputDir.m.
% Last documented: 2026-09-03

thisDir = fileparts(mfilename('fullpath'));
functionsDir = fullfile(thisDir, '..', '..', 'Functions');
if ~exist(functionsDir, 'dir')
    functionsDir = fullfile(thisDir, '..', 'Functions');
end
if exist(functionsDir, 'dir')
    addpath(functionsDir);
else
    error('Shared Functions folder not found from %s.', thisDir);
end

% Own R-022 folder that collects every manuscript figure this driver produces.
manuscriptDir = getFigureOutputDir('MakeManuscriptFigures');
figuresMatlabRoot = fileparts(manuscriptDir);

% Heavy all-cell coverage stages (ExtractAgeingData, PlotEISData) run LAST so the
% single manuscript figures are produced and collected first during an overnight run.
stageNames = { ...
    'characterization_combined (PlotCharacterizationData_Cell15)'; ...
    'ReferencePerformanceCycle (PlotReferencePerformanceCycleFigure_Cell68)'; ...
    'LiStrippingMethods_a/_b (PlotLiStrippingMethods_Cell35)'; ...
    'OCP Anode/Cathode (extractOCPLines)'; ...
    'CharacterizationResults Cell_4'; ...
    'CharacterizationResults Cell_6'; ...
    'CharacterizationResults Cell_15'; ...
    'Calendar/CyclicAgeing (PlotAgeingCombinedOverview)'; ...
    'Cell summary Cell_35 (PlotCellSummary)'; ...
    'All cyclic/calendar cell overviews (ExtractAgeingData)'; ...
    'All ageing EIS diagrams (PlotEISData)'};

% Generator output stem for each stage; '' skips consolidation for the heavy
% all-cell coverage stages (they write hundreds of per-cell diagnostics, not
% single manuscript figures).
stageStems = { ...
    'PlotCharacterizationData_Cell15'; ...
    'PlotReferencePerformanceCycleFigure_Cell68'; ...
    'PlotLiStrippingMethods_Cell35'; ...
    'extractOCPLines'; ...
    'PlotCharacterizationResults'; ...
    'PlotCharacterizationResults'; ...
    'PlotCharacterizationResults'; ...
    'PlotAgeingCombinedOverview'; ...
    'PlotCellSummary'; ...
    ''; ...
    ''};

nStage = numel(stageNames);
stageOk = false(nStage, 1);
for stageIdx = 1:nStage
    fprintf('\n[MakeManuscriptFigures] === Stage %d/%d: %s ===\n', ...
        stageIdx, nStage, stageNames{stageIdx});
    tStageStart = now; %#ok<TNOW1> % serial date number, matches dir().datenum
    try
        switch stageIdx
            case 1, runStage(thisDir, 'PlotCharacterizationData_Cell15.m');
            case 2, runStage(thisDir, 'PlotReferencePerformanceCycleFigure_Cell68.m');
            case 3, runStage(thisDir, 'PlotLiStrippingMethods_Cell35.m');
            case 4, runStage(thisDir, 'extractOCPLines.m');
            case 5, runCharacterizationResults(thisDir, 'Cell_4');
            case 6, runCharacterizationResults(thisDir, 'Cell_6');
            case 7, runCharacterizationResults(thisDir, 'Cell_15');
            case 8, runStage(thisDir, 'PlotAgeingCombinedOverview.m');
            case 9, runCellSummary(thisDir, 'Cell_35');
            case 10, runAllCellOverviews(thisDir);
            case 11, runAllCellEis(thisDir);
        end
        stageOk(stageIdx) = true;
        if ~isempty(stageStems{stageIdx})
            consolidateStage(manuscriptDir, figuresMatlabRoot, stageStems{stageIdx}, tStageStart);
        end
        fprintf('[MakeManuscriptFigures] PASS: %s\n', stageNames{stageIdx});
    catch err
        fprintf(2, '[MakeManuscriptFigures] FAIL: %s\n    %s\n', ...
            stageNames{stageIdx}, err.message);
    end
    close all;
    % Recover driver state clobbered by the generator scripts' 'clear' calls:
    % runStage isolates them in a function workspace, so nothing to restore.
end

fprintf('\n[MakeManuscriptFigures] ======= Summary =======\n');
for stageIdx = 1:nStage
    if stageOk(stageIdx); status = 'PASS'; else; status = 'FAIL'; end
    fprintf('  %-4s %s\n', status, stageNames{stageIdx});
end
fprintf('[MakeManuscriptFigures] %d/%d stages passed. Manuscript figures collected in %s\n', ...
    sum(stageOk), nStage, manuscriptDir);

%% Local functions (each gives the generator script a disposable workspace)
function consolidateStage(manuscriptDir, figuresRoot, stem, tStart)
% Copy the PDF/PNG files the stage's generator wrote (mtime >= tStart) into the
% MakeManuscriptFigures folder so every manuscript figure lives in one place.
srcDir = fullfile(figuresRoot, stem);
if ~exist(srcDir, 'dir'); return; end
files = [dir(fullfile(srcDir, '*.pdf')); dir(fullfile(srcDir, '*.png'))];
for k = 1:numel(files)
    if files(k).datenum + 1e-6 >= tStart
        try
            copyfile(fullfile(srcDir, files(k).name), fullfile(manuscriptDir, files(k).name));
        catch copyErr
            fprintf(2, '[MakeManuscriptFigures] WARN copy %s: %s\n', ...
                files(k).name, copyErr.message);
        end
    end
end
end

function runStage(thisDir, scriptName)
run(fullfile(thisDir, scriptName));
end

function runCharacterizationResults(thisDir, cellId)
% Runs PlotCharacterizationResults.m for one characterization cell via its
% cellNumOverride hook (outputs are cell-tagged, so runs do not overwrite).
cellNumOverride = cellId;  %#ok<NASGU> % consumed by the script's guard block
run(fullfile(thisDir, 'PlotCharacterizationResults.m'));
end

function runCellSummary(thisDir, cellId)
% Run the manuscript's selected example-cell summary via the supported override.
cellNumOverride = cellId; %#ok<NASGU> % consumed by PlotCellSummary.m
run(fullfile(thisDir, 'PlotCellSummary.m'));
end

function runAllCellOverviews(thisDir)
% Regenerate each cyclic and calendar cell's overview diagnostics in the
% ExtractAgeingData R-022 folder without writing to the read-only data tree.
dataRoot = '\\tsn.tno.nl\RA-Data\SV\sv-072952\BTS Data\NEXTBMS\ZenodoRoot';
cyclicDir = fullfile(dataRoot, '4_Ageing', 'Cyclic_ageing_data');
calendarDir = fullfile(dataRoot, '4_Ageing', 'Calendar_ageing_data');
sourceFolders = {cyclicDir, calendarDir};
sourceLabels = {'cyclic', 'calendar'};
failures = {};

% Enumerate the actual dataset folders so Wave 8 covers every available cell.
for sourceIndex = 1:numel(sourceFolders)
    cellFolders = dir(fullfile(sourceFolders{sourceIndex}, 'Cell_*'));
    cellFolders = cellFolders([cellFolders.isdir]);
    for cellIndex = 1:numel(cellFolders)
        cellId = cellFolders(cellIndex).name;
        fprintf('[MakeManuscriptFigures] Overview %s %d/%d: %s\n', ...
            sourceLabels{sourceIndex}, cellIndex, numel(cellFolders), cellId);
        try
            runExtractAgeingData(thisDir, cellId, sourceFolders{sourceIndex});
            fprintf('[MakeManuscriptFigures] PASS overview: %s\n', cellId);
        catch err
            failures{end+1, 1} = sprintf('%s (%s): %s', ...
                cellId, sourceLabels{sourceIndex}, err.message); %#ok<AGROW>
            fprintf(2, '[MakeManuscriptFigures] FAIL overview: %s\n    %s\n', ...
                cellId, err.message);
        end
        close all;
    end
end

% Finish the full roster before surfacing failures to the enclosing stage.
if ~isempty(failures)
    error('MakeManuscriptFigures:AllCellOverviewsFailed', ...
        '%d all-cell overview run(s) failed:\n%s', numel(failures), strjoin(failures, '\n'));
end
end

function runExtractAgeingData(thisDir, cellId, sourceFolder)
% Supply the entry script's supported overrides from an isolated workspace.
cellNumOverride = cellId; %#ok<NASGU> % consumed by ExtractAgeingData.m
folderOverride = sourceFolder; %#ok<NASGU> % consumed by ExtractAgeingData.m
run(fullfile(thisDir, 'ExtractAgeingData.m'));
end

function runAllCellEis(thisDir)
% Regenerate every EIS input file from every available ageing life-stage folder.
dataRoot = '\\tsn.tno.nl\RA-Data\SV\sv-072952\BTS Data\NEXTBMS\ZenodoRoot';
eisRootDir = fullfile(dataRoot, '4_Ageing', 'EIS_data');
stageFolders = dir(fullfile(eisRootDir, '*_EIS'));
stageFolders = stageFolders([stageFolders.isdir]);
eisOutputDir = getFigureOutputDir('PlotEISData');
failures = {};

% One PlotEISData invocation per stage processes all of that stage's cell files.
for stageIndex = 1:numel(stageFolders)
    stageName = stageFolders(stageIndex).name;
    stageDir = fullfile(stageFolders(stageIndex).folder, stageName);
    inputFiles = dir(fullfile(stageDir, '*_impedanceData.csv'));
    fprintf('[MakeManuscriptFigures] EIS stage %d/%d: %s (%d files)\n', ...
        stageIndex, numel(stageFolders), stageName, numel(inputFiles));
    try
        runEisStage(thisDir, stageDir);

        % Confirm every source file produced a unique, stage-qualified R-022 PNG.
        for fileIndex = 1:numel(inputFiles)
            [~, fileBaseName] = fileparts(inputFiles(fileIndex).name);
            outputFile = fullfile(eisOutputDir, [stageName '_' fileBaseName '_NyquistPlot.png']);
            if ~exist(outputFile, 'file')
                error('MakeManuscriptFigures:MissingEisOutput', ...
                    'Expected EIS output was not created: %s', outputFile);
            end
        end
        fprintf('[MakeManuscriptFigures] PASS EIS stage: %s\n', stageName);
    catch err
        failures{end+1, 1} = sprintf('%s: %s', stageName, err.message); %#ok<AGROW>
        fprintf(2, '[MakeManuscriptFigures] FAIL EIS stage: %s\n    %s\n', ...
            stageName, err.message);
    end
    close all;
end

% Finish all stages before reporting any failed stage to the enclosing driver.
if ~isempty(failures)
    error('MakeManuscriptFigures:AllCellEisFailed', ...
        '%d EIS stage run(s) failed:\n%s', numel(failures), strjoin(failures, '\n'));
end
end

function runEisStage(thisDir, stageDir)
% Supply the supported stage override from an isolated workspace.
eisFolderOverride = stageDir; %#ok<NASGU> % consumed by PlotEISData.m
run(fullfile(thisDir, 'PlotEISData.m'));
end
